File: src/test/java/net/sjrx/intellij/plugins/systemdunitfiles/documentation/UnitFileDocumentationProviderTest.java
package net.sjrx.intellij.plugins.systemdunitfiles.documentation;

import com.intellij.lang.ASTNode;
import com.intellij.lang.Language;
import com.intellij.lang.documentation.DocumentationProviderEx;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Key;
import com.intellij.openapi.util.NlsSafe;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiInvalidElementAccessException;
import com.intellij.psi.PsiManager;
import com.intellij.psi.PsiReference;
import com.intellij.psi.ResolveState;
import com.intellij.psi.impl.FakePsiElement;
import com.intellij.psi.scope.PsiScopeProcessor;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.search.SearchScope;
import com.intellij.util.IncorrectOperationException;
import net.sjrx.intellij.plugins.systemdunitfiles.AbstractUnitFileTest;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Icon;
import java.util.Collections;
import java.util.List;

public class UnitFileDocumentationProviderTest extends AbstractUnitFileTest {


  private final DocumentationProviderEx sut = new UnitFileDocumentationProvider();

  public void testGetUrlForUnknownKeyReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "BadKey=Hello Good Sir";

    PsiFile psiFile = setupFileInEditor("file.service", file);


    // Exercise SUT
    PsiElement badKey = getAllKeysInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(badKey, badKey);

    // Verification
    assertNull(urls);
  }

  public void testGetUrlForUnknownKeySeparatorReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "BadKey=Hello Good Sir";

    PsiFile psiFile = setupFileInEditor("file.service", file);


    // Exercise SUT
    PsiElement badKeySeparator = getAllSeparatorsInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(badKeySeparator, badKeySeparator);

    // Verification
    assertNull(urls);
  }

  public void testGetUrlForKnownKeyReturnsAValidUrl() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement documentationKey = getAllKeysInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(documentationKey, documentationKey);

    // Verification
    assertEquals(Collections.singletonList("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Documentation="), urls);
  }

  public void testGetUrlForKnownKeySeparatorReturnsAValidUrl() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement documentationKeySeparator = getAllSeparatorsInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(documentationKeySeparator, documentationKeySeparator);

    // Verification
    assertEquals(Collections.singletonList("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Documentation="), urls);
  }
  
  public void testGetUrlForDeprecatedOptionReturnsAValidUrl() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "BlockIOAccounting=true";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationKeySeparator = getAllSeparatorsInFile(psiFile).get(0);
  
    List<String> urls = sut.getUrlFor(documentationKeySeparator, documentationKeySeparator);
    
    // Verification
    assertEquals(Collections.singletonList("https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#BlockIOAccounting="), urls);
  }

  public void testGetUrlForUnknownSectionReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    String file = "[X-Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement unknownSectionHeader = getAllSectionInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(unknownSectionHeader, unknownSectionHeader);

    // Verification
    assertNull(urls);
  }

  public void testGetUrlForKnownSectionReturnsAValidUrl() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement sectionHeader = getAllSectionInFile(psiFile).get(0);

    List<String> urls = sut.getUrlFor(sectionHeader, sectionHeader);

    // Verification
    assertEquals(Collections.singletonList("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#%5BUnit%5D%20Section%20Options"), urls);
  }


  public void testGenerateDocForUnknownKeyReturnsNull() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "BadKey=Hello Good Sir";

    PsiFile psiFile = setupFileInEditor("file.service", file);


    // Exercise SUT
    PsiElement badKey = getAllKeysInFile(psiFile).get(0);

    String doc = sut.generateDoc(badKey, badKey);

    // Verification
    assertNull(doc);
  }

  public void testGenerateDocForUnknownKeySeparatorReturnsNull() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "BadKey=Hello Good Sir";

    PsiFile psiFile = setupFileInEditor("file.service", file);


    // Exercise SUT
    PsiElement badKeySeparator = getAllSeparatorsInFile(psiFile).get(0);

    String doc = sut.generateDoc(badKeySeparator, badKeySeparator);

    // Verification
    assertNull(doc);
  }

  public void testGenerateDocForKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement documentationKey = getAllKeysInFile(psiFile).get(0);

    String doc = sut.generateDoc(documentationKey, documentationKey);

    // Verification
    assertNotNull(doc);
    assertTrue((doc.length() > 0));
  }

  public void testGenerateDocForKnownKeySeparatorReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement documentationKeySeparator = getAllSeparatorsInFile(psiFile).get(0);

    String doc = sut.generateDoc(documentationKeySeparator, documentationKeySeparator);

    // Verification
    assertNotNull(doc);
    assertTrue((doc.length() > 0));
  }

  public void testGenerateDocForUnknownSectionReturnsNull() {
    // Fixture Setup
    String file = "[X-Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement unknownSectionHeader = getAllSectionInFile(psiFile).get(0);

    String doc = sut.generateDoc(unknownSectionHeader, unknownSectionHeader);

    // Verification
    assertNull(doc);
  }

  public void testGenerateDocForKnownSectionReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";

    PsiFile psiFile = setupFileInEditor("file.service", file);

    // Exercise SUT
    PsiElement unknownSectionHeader = getAllSectionInFile(psiFile).get(0);

    String doc = sut.generateDoc(unknownSectionHeader, unknownSectionHeader);

    // Verification
    assertNotNull(doc);

    assertTrue(doc.length() > 1);

  }
  
  public void testGenerateDocForCompletedValueOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllCompletedValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNotNull(doc);
    assertTrue((doc.length() > 0));
  }
  
  public void testGenerateDocForCompletedValueOfUnknownSectionReturnsNull() {
    // Fixture Setup
    String file = "[X-Unit]\n"
                  + "Documentation=http://www.google.com";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllCompletedValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }
  
  
  public void testGenerateDocForCompletedValueOfUnknownKeyReturnsNull() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "HelpPage=http://www.google.com";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllCompletedValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }
  
  public void testGenerateDocForContinuingValueOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com\\\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
  
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNotNull(doc);
    assertTrue((doc.length() > 0));
  }
  
  public void testGenerateDocForContinuingValueOfUnknownSectionReturnsNull() {
    // Fixture Setup
    String file = "[X-Unit]\n"
                  + "Documentation=http://www.google.com\\\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
  
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }
  
  
  public void testGenerateDocForContinuingValueOfUnknownKeyReturnsNull() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "HelpPage=http://www.google.com\\\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
  
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }
  
  public void testGenerateDocForContinuingValueAfterCommentsOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "Documentation=http://www.google.com\\\n"
                  + ";Good bye\n"
                  + "#See ya later!\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNotNull(doc);
    assertTrue((doc.length() > 0));
  }
  
  public void testGenerateDocForContinuingValueAfterCommentsOfUnknownSectionReturnsNull() {
    // Fixture Setup
    String file = "[X-Unit]\n"
                  + "Documentation=http://www.google.com\\\n"
                  + ";Good bye\n"
                  + "#See ya later!\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }
  
  
  public void testGenerateDocForContinuingValueAfterCommentsOfUnknownKeyReturnsNull() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "HelpPage=http://www.google.com\\\n"
                  + ";Good bye\n"
                  + "#See ya later!\n"
                  + "hello sir";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement documentationValue = getAllContinuingValuesInFile(psiFile).get(0);
    
    documentationValue = sut.getCustomDocumentationElement(myFixture.getEditor(), psiFile, documentationValue, 0);
    
    String doc = sut.generateDoc(documentationValue, documentationValue);
    
    // Verification
    assertNull(doc);
  }

  public void testGenerateDocMovedSectionReturnsSomeText() {
    // Fixture Setup
    String file = "[Unit]\n"
                  + "PropagateReloadTo=cpu.mount";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement renamedSection = getAllKeysInFile(psiFile).get(0);
    
    String doc = sut.generateDoc(renamedSection, renamedSection);
    
    // Verification
    assertNotNull(doc);
    
    assertEquals("Expected the generated documentation to match", "<div class='definition'><pre>PropagateReloadTo</pre>"
                                                                  + "</div><div class='content'><p>The key <var>PropagateReloadTo</var> in"
                                                                  + " section <b>Unit</b> has been renamed to <var>PropagateReloadsTo</var>"
                                                                  + "<p>NOTE: The semantics of the new value may not"
                                                                  + " match the existing value.<p>"
                                                                  + "<a href='https://github.com/systemd/systemd/commit/7f2cddae09'>More"
                                                                  + " information is available here</a></div>", doc);
    
  }
  
  public void testGenerateDocForUnsupportedSectionReturnsSomeValidText() {
    // Fixture Setup
    String file = "[Service]\n"
                  + "PermissionsStartOnly=true";
    
    PsiFile psiFile = setupFileInEditor("file.service", file);
    
    // Exercise SUT
    PsiElement unsupportedSection = getAllKeysInFile(psiFile).get(0);
    
    String doc = sut.generateDoc(unsupportedSection, unsupportedSection);
    
    // Verification
    assertNotNull(doc);
  
    assertEquals("Expected the generated documentation to match", "<div class='definition'><pre>PermissionsStartOnly"
                                                                  + "</pre></div><div class='content'><p><var>PermissionsStartOnly</var> in"
                                                                  + " section <b>Service</b> is not officially supported.<p>"
                                                                  + "<a href='https://github.com/systemd/systemd/blob/v241/NEWS#L561'>"
                                                                  + "More information is available here</a></div>", doc);
  }
  
  public void testGenerateDocFromPsiElementWithNoNodeDoesNotCrash() {
  
    // Fixture Setup
  
    /*
     * The important part of this is that FakePsiElement getNode() returns null.
     */
    PsiElement element = new FakePsiElement() {
      @Override
      public PsiElement getParent() {
        return null;
      }
    };
    
    // Execute SUT
    sut.generateDoc(element, element);
    
    // Verification (Implicit we want to avoid an NPE, so if we don't crash we didn't throw an NPE)
  }
  
  public void testGetUrlFromPsiElementWithNoNodeDoesNotCrash() {
   
    // Fixture Setup
  
    /*
     * The important part of this is that FakePsiElement getNode() returns null.
     */
    PsiElement element = new FakePsiElement() {
      @Override
      public PsiElement getParent() {
        return null;
      }
    };
    
    // Execute SUT
    sut.getUrlFor(element, element);
    
    // Verification (Implicit we want to avoid an NPE, so if we don't crash we didn't throw an NPE).
  }
  
}

File: src/test/kotlin/net/sjrx/intellij/plugins/systemdunitfiles/documentation/UnitFileDocumentationProviderTest.kt
package net.sjrx.intellij.plugins.systemdunitfiles.documentation

import com.intellij.lang.documentation.DocumentationProviderEx
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.FakePsiElement
import junit.framework.TestCase
import net.sjrx.intellij.plugins.systemdunitfiles.AbstractUnitFileTest

class UnitFileDocumentationProviderTest : AbstractUnitFileTest() {
  private val sut: DocumentationProviderEx = UnitFileDocumentationProvider()
  fun testGetUrlForUnknownKeyReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    val file = """
           [Unit]
           BadKey=Hello Good Sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)


    // Exercise SUT
    val badKey = getAllKeysInFile(psiFile)[0]
    val urls = sut.getUrlFor(badKey, badKey)

    // Verification
    TestCase.assertNull(urls)
  }

  fun testGetUrlForUnknownKeySeparatorReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    val file = """
           [Unit]
           BadKey=Hello Good Sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)


    // Exercise SUT
    val badKeySeparator = getAllSeparatorsInFile(psiFile)[0]
    val urls = sut.getUrlFor(badKeySeparator, badKeySeparator)

    // Verification
    TestCase.assertNull(urls)
  }

  fun testGetUrlForKnownKeyReturnsAValidUrl() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val documentationKey = getAllKeysInFile(psiFile)[0]
    val urls = sut.getUrlFor(documentationKey, documentationKey)

    // Verification
    TestCase.assertEquals(listOf("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Documentation="), urls)
  }

  fun testGetUrlForKnownKeySeparatorReturnsAValidUrl() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val documentationKeySeparator = getAllSeparatorsInFile(psiFile)[0]
    val urls = sut.getUrlFor(documentationKeySeparator, documentationKeySeparator)

    // Verification
    TestCase.assertEquals(listOf("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Documentation="), urls)
  }

  fun testGetUrlForDeprecatedOptionReturnsAValidUrl() {
    // Fixture Setup
    val file = """
           [Service]
           BlockIOAccounting=true
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val documentationKeySeparator = getAllSeparatorsInFile(psiFile)[0]
    val urls = sut.getUrlFor(documentationKeySeparator, documentationKeySeparator)

    // Verification
    TestCase.assertEquals(listOf("https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#BlockIOAccounting="), urls)
  }

  fun testGetUrlForUnknownSectionReturnsAnEmptyListOfUrls() {
    // Fixture Setup
    val file = """
           [X-Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val unknownSectionHeader = getAllSectionInFile(psiFile)[0]
    val urls = sut.getUrlFor(unknownSectionHeader, unknownSectionHeader)

    // Verification
    TestCase.assertNull(urls)
  }

  fun testGetUrlForKnownSectionReturnsAValidUrl() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val sectionHeader = getAllSectionInFile(psiFile)[0]
    val urls = sut.getUrlFor(sectionHeader, sectionHeader)

    // Verification
    TestCase.assertEquals(listOf("https://www.freedesktop.org/software/systemd/man/systemd.unit.html#%5BUnit%5D%20Section%20Options"), urls)
  }

  fun testGenerateDocForUnknownKeyReturnsNull() {
    // Fixture Setup
    val file = """
           [Unit]
           BadKey=Hello Good Sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)


    // Exercise SUT
    val badKey = getAllKeysInFile(psiFile)[0]
    val doc = sut.generateDoc(badKey, badKey)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForUnknownKeySeparatorReturnsNull() {
    // Fixture Setup
    val file = """
           [Unit]
           BadKey=Hello Good Sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)


    // Exercise SUT
    val badKeySeparator = getAllSeparatorsInFile(psiFile)[0]
    val doc = sut.generateDoc(badKeySeparator, badKeySeparator)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val documentationKey = getAllKeysInFile(psiFile)[0]
    val doc = sut.generateDoc(documentationKey, documentationKey)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 0)
  }

  fun testGenerateDocForKnownKeySeparatorReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val documentationKeySeparator = getAllSeparatorsInFile(psiFile)[0]
    val doc = sut.generateDoc(documentationKeySeparator, documentationKeySeparator)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 0)
  }

  fun testGenerateDocForUnknownSectionReturnsNull() {
    // Fixture Setup
    val file = """
           [X-Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val unknownSectionHeader = getAllSectionInFile(psiFile)[0]
    val doc = sut.generateDoc(unknownSectionHeader, unknownSectionHeader)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForKnownSectionReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val unknownSectionHeader = getAllSectionInFile(psiFile)[0]
    val doc = sut.generateDoc(unknownSectionHeader, unknownSectionHeader)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 1)
  }

  fun testGenerateDocForCompletedValueOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllCompletedValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 0)
  }

  fun testGenerateDocForCompletedValueOfUnknownSectionReturnsNull() {
    // Fixture Setup
    val file = """
           [X-Unit]
           Documentation=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllCompletedValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForCompletedValueOfUnknownKeyReturnsNull() {
    // Fixture Setup
    val file = """
           [Unit]
           HelpPage=http://www.google.com
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllCompletedValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForContinuingValueOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com\
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 0)
  }

  fun testGenerateDocForContinuingValueOfUnknownSectionReturnsNull() {
    // Fixture Setup
    val file = """
           [X-Unit]
           Documentation=http://www.google.com\
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForContinuingValueOfUnknownKeyReturnsNull() {
    // Fixture Setup
    val file = """
           [Unit]
           HelpPage=http://www.google.com\
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForContinuingValueAfterCommentsOfKnownKeyReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Unit]
           Documentation=http://www.google.com\
           ;Good bye
           #See ya later!
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertTrue(doc!!.length > 0)
  }

  fun testGenerateDocForContinuingValueAfterCommentsOfUnknownSectionReturnsNull() {
    // Fixture Setup
    val file = """
           [X-Unit]
           Documentation=http://www.google.com\
           ;Good bye
           #See ya later!
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocForContinuingValueAfterCommentsOfUnknownKeyReturnsNull() {
    // Fixture Setup
    val file = """
           [Unit]
           HelpPage=http://www.google.com\
           ;Good bye
           #See ya later!
           hello sir
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    var documentationValue = getAllContinuingValuesInFile(psiFile)[0]
    documentationValue = sut.getCustomDocumentationElement(myFixture.editor, psiFile, documentationValue, 0)!!
    val doc = sut.generateDoc(documentationValue, documentationValue)

    // Verification
    TestCase.assertNull(doc)
  }

  fun testGenerateDocMovedSectionReturnsSomeText() {
    // Fixture Setup
    val file = """
           [Unit]
           PropagateReloadTo=cpu.mount
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val renamedSection = getAllKeysInFile(psiFile)[0]
    val doc = sut.generateDoc(renamedSection, renamedSection)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertEquals(
      "Expected the generated documentation to match", "<div class='definition'><pre>PropagateReloadTo</pre>"
        + "</div><div class='content'><p>The key <var>PropagateReloadTo</var> in"
        + " section <b>Unit</b> has been renamed to <var>PropagateReloadsTo</var>"
        + "<p>NOTE: The semantics of the new value may not"
        + " match the existing value.<p>"
        + "<a href='https://github.com/systemd/systemd/commit/7f2cddae09'>More"
        + " information is available here</a></div>", doc
    )
  }

  fun testGenerateDocForUnsupportedSectionReturnsSomeValidText() {
    // Fixture Setup
    val file = """
           [Service]
           PermissionsStartOnly=true
           """.trimIndent()
    val psiFile = setupFileInEditor("file.service", file)

    // Exercise SUT
    val unsupportedSection = getAllKeysInFile(psiFile)[0]
    val doc = sut.generateDoc(unsupportedSection, unsupportedSection)

    // Verification
    TestCase.assertNotNull(doc)
    TestCase.assertEquals(
      "Expected the generated documentation to match", "<div class='definition'><pre>PermissionsStartOnly"
        + "</pre></div><div class='content'><p><var>PermissionsStartOnly</var> in"
        + " section <b>Service</b> is not officially supported.<p>"
        + "<a href='https://github.com/systemd/systemd/blob/v241/NEWS#L561'>"
        + "More information is available here</a></div>", doc
    )
  }

}
